### ISSUE TITLE
[hibernate-search] Support injecting runtime properties in custom `IndexLayoutStrategy` implementation

### ISSUE THREAD

**Issue by cthiebault** [quarkusio/quarkus#41429]
### Description

Following https://github.com/quarkusio/quarkus/issues/41384, I try to implement a custom `IndexLayoutStrategy` to prefix the index name with a config property defined in `application.properties`.

```java
@SearchExtension
public class PrefixIndexLayoutStrategy implements IndexLayoutStrategy {

    @ConfigProperty(name = "index-prefix")
    String prefix;

    @Override
    public String createInitialElasticsearchIndexName(String hibernateSearchIndexName) {
        return prefix + hibernateSearchIndexName + "-000001";
    }
    [...]
```

when I run my application (`prod` profile in k8s), I get

```text
A runtime config property value differs from the value that was injected during the static intialization phase:
- the runtime value of 'index-prefix' is [dev-] but the value [null] was injected into myapp.PrefixIndexLayoutStrategy()

If that's intentional then annotate the injected field/parameter with @io.quarkus.runtime.annotations.StaticInitSafe to eliminate the false positive.
```

Quoting @yrodiere from https://github.com/quarkusio/quarkus/issues/41384#issuecomment-2188252197:

> I believe the index layout strategy is used by Hibernate Search at static init, so you can't use runtime properties in there. 
> I'd suggest opening a feature request to make the index layout strategy a runtime component (which makes sense)
> FWIW it'll probably require changes to Hibernate Search itself, since it currently creates index names very (too?) early, 
> for no good reason that I can see (the names aren't used until runtime): 
> https://github.com/hibernate/hibernate-search/blob/0b9e3f0d87250daa800f10223bcc81cdb0da773a/backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchBackendImpl.java#L206

**Comment by yrodiere** [comment]
Thanks.

Note a workaround would be to annotate your injection point with `@StaticInitSafe`, and to build your app separately for each environment:

```
./mvn clean package -Dquarkus.profile=prod
```

```
./mvn clean package -Dquarkus.profile=staging
```

Then each env will use the value you set for that property -- assuming it's set at build time.

**Comment by cthiebault** [comment]
But in that case I would need to publish a different artefact (Docker image in my case) per environment?

As a workaround, I'm using an environment variable to set the prefix... I don't like to mix config in `application.properties` and env variables, but it works!

**Comment by yrodiere** [comment]
>  FWIW it'll probably require changes to Hibernate Search itself, since it currently creates index names very (too?) early,
for no good reason that I can see (the names aren't used until runtime):
https://github.com/hibernate/hibernate-search/blob/0b9e3f0d87250daa800f10223bcc81cdb0da773a/backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/impl/ElasticsearchBackendImpl.java#L206

I opened https://hibernate.atlassian.net/browse/HSEARCH-5190 to make a fix possible in Quarkus in the future.

> But in that case I would need to publish a different artefact (Docker image in my case) per environment?

Yes.

> As a workaround, I'm using an environment variable to set the prefix... I don't like to mix config in `application.properties` and env variables, but it works!

I suspect you're being lucky and this might break in the future, but it's better than nothing :)

**Comment by yrodiere** [comment]
https://hibernate.atlassian.net/browse/HSEARCH-5189 moved the creation of the index layout strategy to "runtime", so I will try fix this in #41359 by moving the config to runtime as well.
